[#menu] 메뉴 로직

daakludens opened the pull request:
가게(store)에서 가지고 있는 메뉴 관련한 로직
- 수정
- 저장
- 삭제

daakludens committed:
[#menu] feat. menu 로직(저장, 수정, 삭제) 생성

Jane096 commented on src/main/java/com/konacafe/controller/MenuController.java:
```diff
@@ -0,0 +1,32 @@
+        menuService.save(requestDto);
+    }
+
+    @PutMapping("/{id}")
+    public void update(@PathVariable Long id, @RequestBody MenuSaveRequestDto requestDto) {
```
`@PathVariable` 은 엔드포인트 노출 위험이 있을 것 같은데 `@RequestParam` 은 어떠신가요?

daakludens replied:
찾아보니 @PathVariable의 경우 URI 경로에서 직접 값을 가져오기 때문에 정보 노출 위험이 있고 인코딩 안 되는 문제가 있네요.  또한 required 속성을 줄 때 URI 경로가 변동되는 문제점도 발생할 수 있고요. 수정하겠습니다.

Jane096 commented on src/main/java/com/konacafe/domain/menu/Menu.java:
```diff
@@ -0,0 +1,33 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
+    private Long id;
+
+    private String name;
```
다른 `Store` 나 기타 여러 테이블에서도 `id`, `name` 은 중복으로 사용될 것 같은데 나중에 혼동이 올 가능성이 있을 것 같으니 `menu_id`, `menu_name` 은 어떨까요?

daakludens replied:
가독성을 높이려 최소한의 정보만 담으려고 했는데 다른 entity 모두 같은 방식으로 쓰이면 확실히 헷갈릴 것 같네요. 팀 개발 차원에서 유지 보수성을 높이기 위해서 어느 entity의 id 또는 name인지 명시하는 게 좋을 것 같습니다.

Jane096 commented on src/main/java/com/konacafe/service/MenuService.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Transactional
+    public void update(Long id, MenuSaveRequestDto requestDto) {
+        Menu menu = menuRepository
+                .findById(id)
+                .orElseThrow(() -> new IllegalArgumentException("해당되는 메뉴가 없습니다."));
```
나중에 로그인 권한 체크하는 로직으로 대체가 될 것 같은데 어떻게 생각하시나요?

daakludens replied:
음.. 로그인 권한 체크 기능도 고려해야 겠지만 메뉴 수정 기능은 다르게 접근해야 할 것 같네요. 수정이란 건 이미 존재하는 메뉴의 정보를 부분 갱신하는 건데 메뉴가 없을리는 없죠. 이건 로직 flaw인 것 같습니다.
차라리 갱신 중 에러 발생 시 exception을 던지게 하는 게 맞는 거 같습니다. 어떻게 생각하시나요?

Jane096 replied:
아하 말씀 주신 내용 이해했습니다. 우선, 갱신 전 사용자가 해당 글을 작성한 사람인지 본인 확인을 한 후, 아니라면 exception을 던지는 것을 생각해 본 적이 있는데 혹시 이것 이외에 생각해보신 exception이 있으신가요?

> validation 체크 같은 것도 포함될 수 있겠네요 :)

daakludens replied:
제가 말씀드린 exception도 사용자 본인 확인 관련 exception이었습니다. 본인 확인 로직이 정해지면 그 후 추가하면 될 것 같네요.

daakludens committed:
[#menu] refactor : 피드백 반영해 변경
- DTO 대신 model 시점으로 파일 및 객체명 변경
- entity의 필드명을 명시적으로 변경
- 안전하게 파라미터를 받아올 수 있도록 @RequestParam으로 변경
- update 관련 로직 우선 보류
